### ISSUE TITLE
SHAP in unsupervised anomaly detection

### ISSUE THREAD

**Issue by asstergi** [shap/shap#33]
Hi @slundberg ,

I'm facing a specific use case where I try to interpret multivariate anomaly detection algorithms. So, I have a dataset, I apply an unsupervised anomaly detection algorithm (e.g. Isolation Forest) and get which data samples are anomalous. Then, I build a supervised model (XGBoost) and I'm getting the SHAP values in order to explain the anomalies.

I have a question that is more or less 'philosophical'. Since I will not be using the supervised model to make predictions on new data, how should the model be trained? Is it ok to train the XGBoost model on the dataset and then make the predictions (and get the SHAP values) on the same dataset? Could the SHAP values 'overfit' since I'm using the model on the dataset it was trained on? Or should I, still, create different train-test datasets? 

My guess is that I should still use different train/test sets but I'm not convinced this is the correct approach. Any thoughts?

**Comment by slundberg** [comment]
Cool application. What you want is an XGBoost model that would do a good job identifying new anomalies in future data. If your model doesn't, then it is not capturing real structure in the data.

So since SHAP values are only as meaningful as the model they explain, I would for sure do a train test split to determine the XGBoost parameters. However, once you have those params I would also then retrain the XGBoost model on the whole dataset for a bit of extra accuracy before explaining it.

**Comment by asstergi** [comment]
Thank you for you comments. I absolutely agree that the XGBoost should be able to capture the real structure of the data.

So, are you suggesting doing cross-validation to get the optimal hyper-parameters and then re-train the model on the whole dataset (and make predictions on the same one)?

**Comment by slundberg** [comment]
That’s right. If you want to know why a model is making predictions I think
it’s fine to use a training set example.

**Comment by jsulopzs** [comment]
Is it possible to output shap values out of an Isolation Forest by its own, without the need to use `xgboost`?

I have tried to compute the shap values for Anomaly Detection with Local Outlier Factor, SVM and Isolation Forest. They all output -1 or 1 with respect to if the observation is an outlier or not, respectively.

My question is whether we need `xgboosting` to calculate probabilities to find outliers and then compute the shap values, or if there is another way of computing the shap values of these methods by its own?

Every time that I have run the experiment with the outcome labelled -1 or 1, I couldn't match the results because they all yielded 1 as if the shap values don't go well with these models...

**Comment by Harshit200107** [comment]
I am using Anomaly Detection with Local Outlier Forest, One Class Svm ,Isolation Forest .They have Output as an outlier 1 or -1.
 
I am trying to use Shap For finding interpretation  of these models but It is saying These Explainer doesnot go well with these models except for isolation forest.
 
Is there any explainer that works with Local Outlier Forest and One Class Svm

**Comment by skngetich** [comment]
@Harshit200107 did you get a solution? if not what did you use as an alternative? I have the same problem

**Comment by smc181002** [comment]
the LOF in sklearn provides a method called [decision_function](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.LocalOutlierFactor.html#sklearn.neighbors.LocalOutlierFactor.decision_function) which outputs the raw scores of the prediction. Using this will give better explanation of the features. But the `novelty` parameter should be set to `True` for this

**Comment by ETTAN93** [comment]
I am using Isolation Forest for an unsupervised problem as well. How do I apply shap values to see which features are important to predicting the anomalies? Is it via shap.TreeExplainer?
